File: fc-chat/demo/part3/ch4/message-system-add-push/src/main/java/net/prostars/messagesystem/handler/websocket/CreateRequestHandler.java
package net.prostars.messagesystem.handler.websocket;

import java.util.List;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import net.prostars.messagesystem.constant.IdKey;
import net.prostars.messagesystem.constant.MessageType;
import net.prostars.messagesystem.constant.ResultType;
import net.prostars.messagesystem.dto.domain.Channel;
import net.prostars.messagesystem.dto.domain.UserId;
import net.prostars.messagesystem.dto.websocket.inbound.CreateRequest;
import net.prostars.messagesystem.dto.websocket.outbound.CreateResponse;
import net.prostars.messagesystem.dto.websocket.outbound.ErrorResponse;
import net.prostars.messagesystem.dto.websocket.outbound.JoinNotification;
import net.prostars.messagesystem.service.ChannelService;
import net.prostars.messagesystem.service.ClientNotificationService;
import net.prostars.messagesystem.service.UserService;
import org.springframework.data.util.Pair;
import org.springframework.stereotype.Component;
import org.springframework.web.socket.WebSocketSession;

@Component
@SuppressWarnings("unused")
public class CreateRequestHandler implements BaseRequestHandler<CreateRequest> {

  private final ChannelService channelService;
  private final UserService userService;
  private final ClientNotificationService clientNotificationService;

  public CreateRequestHandler(
      ChannelService channelService,
      UserService userService,
      ClientNotificationService clientNotificationService) {
    this.channelService = channelService;
    this.userService = userService;
    this.clientNotificationService = clientNotificationService;
  }

  @Override
  public void handleRequest(WebSocketSession senderSession, CreateRequest request) {
    UserId senderUserId = (UserId) senderSession.getAttributes().get(IdKey.USER_ID.getValue());

    List<UserId> participantIds = userService.getUserIds(request.getParticipantUsernames());
    if (participantIds.isEmpty()) {
      clientNotificationService.sendMessage(
          senderSession,
          senderUserId,
          new ErrorResponse(MessageType.CREATE_REQUEST, ResultType.NOT_FOUND.getMessage()));
      return;
    }

    Pair<Optional<Channel>, ResultType> result;
    try {
      result = channelService.create(senderUserId, participantIds, request.getTitle());
    } catch (Exception ex) {
      clientNotificationService.sendMessage(
          senderSession,
          senderUserId,
          new ErrorResponse(MessageType.CREATE_REQUEST, ResultType.FAILED.getMessage()));
      return;
    }

    if (result.getFirst().isEmpty()) {
      clientNotificationService.sendMessage(
          senderSession,
          senderUserId,
          new ErrorResponse(MessageType.CREATE_REQUEST, result.getSecond().getMessage()));
      return;
    }

    Channel channel = result.getFirst().get();
    clientNotificationService.sendMessage(
        senderSession, senderUserId, new CreateResponse(channel.channelId(), channel.title()));
    participantIds.forEach(
        participantId ->
            CompletableFuture.runAsync(
                () ->
                    clientNotificationService.sendMessage(
                        participantId,
                        new JoinNotification(channel.channelId(), channel.title()))));
  }
}

File: fc-chat/demo/part3/with-push/src/main/kotlin/com/yscorp/withpush/messagesystem/handler/websocket/CreateRequestHandler.kt
package com.yscorp.withpush.messagesystem.handler.websocket

import com.yscorp.withpush.messagesystem.constant.IdKey
import com.yscorp.withpush.messagesystem.constant.MessageType
import com.yscorp.withpush.messagesystem.constant.ResultType
import com.yscorp.withpush.messagesystem.dto.domain.Channel
import com.yscorp.withpush.messagesystem.dto.domain.UserId
import com.yscorp.withpush.messagesystem.dto.websocket.inbound.CreateRequest
import com.yscorp.withpush.messagesystem.dto.websocket.outbound.CreateResponse
import com.yscorp.withpush.messagesystem.dto.websocket.outbound.ErrorResponse
import com.yscorp.withpush.messagesystem.dto.websocket.outbound.JoinNotification
import com.yscorp.withpush.messagesystem.service.ChannelService
import com.yscorp.withpush.messagesystem.service.ClientNotificationService
import com.yscorp.withpush.messagesystem.service.UserService
import org.springframework.stereotype.Component
import org.springframework.web.socket.WebSocketSession
import java.util.concurrent.CompletableFuture
import java.util.function.Consumer

@Component
@Suppress("unused")
class CreateRequestHandler(
    private val channelService: ChannelService,
    private val userService: UserService,
    private val clientNotificationService: ClientNotificationService
) : BaseRequestHandler<CreateRequest> {

    override fun handleRequest(senderSession: WebSocketSession, request: CreateRequest) {
        val senderUserId: UserId = senderSession.attributes[IdKey.USER_ID.value] as UserId

        val participantIds: List<UserId> = userService.getUserIds(request.participantUsernames)
        if (participantIds.isEmpty()) {
            clientNotificationService.sendMessage(
                senderSession,
                senderUserId,
                ErrorResponse(MessageType.CREATE_REQUEST, ResultType.NOT_FOUND.message)
            )
            return
        }

        val result: Pair<Channel?, ResultType>

        try {
            result = channelService.create(senderUserId, participantIds, request.title)
        } catch (ex: Exception) {
            clientNotificationService.sendMessage(
                senderSession,
                senderUserId,
                ErrorResponse(MessageType.CREATE_REQUEST, ResultType.FAILED.message)
            )
            return
        }

        if (result.first == null) {
            clientNotificationService.sendMessage(
                senderSession,
                senderUserId,
                ErrorResponse(MessageType.CREATE_REQUEST, result.second.message)
            )
            return
        }

        val channel: Channel = result.first!!

        clientNotificationService.sendMessage(
            senderSession, senderUserId, CreateResponse(channel.channelId, channel.title)
        )

        participantIds.forEach(
            Consumer { participantId: UserId? ->
                CompletableFuture.runAsync {
                    clientNotificationService.sendMessage(
                        participantId,
                        JoinNotification(channel.channelId, channel.title)
                    )
                }
            })
    }
}
